graf2d/gpadv7/inc/ROOT/RAttrEnum.hxx

```diff
@@ -0,0 +1,67 @@
+/*************************************************************************
+ * Copyright (C) 1995-2021, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RAttrEnum
+#define ROOT7_RAttrEnum
+
+#include <ROOT/RAttrValue.hxx>
+
+namespace ROOT {
+namespace Experimental {
+
+/** \class RAttrValue
+\ingroup GpadROOT7
+\author Sergey Linev <s.linev@gsi.de>
+\date 2021-06-21
+\brief Template class to access single enum value from drawable or other attributes aggregations
+\warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback is welcome!
+*/
+
+#ifdef __CLING__
+template<typename T>
+#else
+template<typename T, typename std::enable_if<std::is_enum<T>::value>::type* = nullptr>
+#endif
+class RAttrEnum : public RAttrValue<T> {
+
+protected:
+
+   T fMaximum;                ///<!  maximal value
```

Comment:
If `enum` can be `OR`-ed - one can just use `RAttrValue<EStyle>`.  It is just `int` <-> `EStyle` casting.
Need of specialized `RAttrEnum<EStyle>`:
1. Check for allowed ranges (0..max)
2. I plan to provide (optionally) string map for all allowed values. In this case not plain integer but string values can be used in CSS.

In such case one can have CSS:
```css
marker { marker_style: dot; }
```

Reply:
OK thanks - as long as we don't have functions that take `int` we're safe. I'm still not "sold" on checking the max - esp for OR-ed value this is error-prone. And why would 0 be the minimum? ;-)

Reply:
> I'm still not "sold" on checking the max - esp for OR-ed value this is error-prone. 

I do not plan to use `RAttrEnum` for OR-ed values and maximum is not applied there.

> And why would 0 be the minimum? ;-)

Just while I plan to use it for very few cases like line_style, fill_style, marker_style and few mores.
There we have under our control there to start enum. 
And extra 2 entries in dictionary should not be a big deal.
For all other cases one could use `RAttrValue<int>`

Reply:
>  For all other cases one could use RAttrValue<int>

Aren't we losing type safety then? I think this is why we need `RAttrEnum<E>` - to avoid assigning a constant from the wrong `enum` to the `int`, like `attr.fillStyle = RLine::EStyle::kDashed`.

Reply:
No, such assignment will not be allowed. It is not possible to simply assign int or other enum.

Only when reading value from container, automatical cast from `int` to `enum` is done - without any checks if value valid or not. Only in `RAttrValue` such check introduced. And as I said - I want add more customization here in the future

Reply:
> I do not plan to use RAttrEnum for OR-ed values and maximum is not applied there.

See e.g. https://wiggling-bits.net/using-enum-classes-as-type-safe-bitmasks/ - there are plenty of these examples out there. You can get type-safety *and* bit operators, and given how often these enum constants will be used this might be useful for us.

Reply:
Again, `RAttrValue<EType>` will provide type safety.
You will not be able to simply assign `int` to that value

Reply:
I'm lost - `RAttrValue` was limited to very few specializations, intentionally, now it becomes an open set of types? In your approach, if you OR constants you lose the type. That seems to contradict "You will not be able to simply assign `int` to that value"?

In my mind we need:
- enum classes
- `RAttrEnum` that simply casts from the enum type to an underlying `RAttrEnum<int>`
- custom operators

Reply:
> I'm lost - RAttrValue was limited to very few specializations, intentionally, now it becomes an open set of types? 

There are 6 types supported: `bool`, `int`, `double`, `std::string`, `RColor`, `RPadLength`.
All other classes will fail to compile. 
```cpp
template <typename Q, bool = std::is_enum<Q>::value>
   struct ValueExtractor {
      Q Get(const RAttrMap::Value_t *value)
      {
         return (Q) RAttrMap::Value_t::GetValue<int>(value);
      }
   };
```
With the code we allow to use any `enum` - but only `enum` as integer.
API only allows to assign `enum` type - any attempt assign `int` refused by compiler.
If one use bit-wise operators for same `enum` type - it may compile but may not.
And one need to cast anyway like:
```
rect.border.style = (RAttrLine::EStyle) RAttrLine::kSolid | RAttrLine::kDashed;
```
But this is illegal syntax anyway. One only should do:
```
rect.border.style = RAttrBorder::kSolid;
```
Here example why I want to have `EStyle` invisible. `RAttrBorder` derived from `RAttrLine`. 
But if I want to assign style for the border, I still can use `RAttrBorder` prefix.